src/spc/codegen.jl

```diff
@@ -0,0 +1,486 @@
+            else
+                trancall = Expr(:call, fn, (cg_expr!(state, v) for v in val.values)...)
+            end
+            kw = Expr(:kw, :tran, trancall)
+            push!(kws, kw)
+        else
+            @show val
+            error("unhandled voltage value $(String(val))")
+        end
+    end
+    return cg_spice_instance!(state, [instance.pos, instance.neg], instance.name, constructor, kws)
+end
+
+function devtype_param(model_kind, mosfet_kind)
+    if model_kind.name === :bsim4
+        return :TYPE => (mosfet_kind == :pmos ? -1 : 1)
+    elseif startswith(String(model_kind.name), "bsimcmg")
+        return :DEVTYPE => (mosfet_kind == :pmos ? 0 : 1)
+    elseif model_kind == :UnimplementedDevice
+        # skip
+        return nothing
+    else
+        error("Needs to be filled in per model")
+    end
+end
+
+function cg_model_def!(state::CodegenState, (model, modelref)::Pair{<:SNode, GlobalRef}, bins::Dict{Symbol, Vector{Symbol}})
+    params = Any[]
+    typ = LSymbol(model.typ)
+    mosfet_type = typ in (:nmos, :pmos) ? typ : nothing
+    level = nothing
+    version = nothing
+    for p in model.parameters
+        name = LSymbol(p.name)
+        if name == :type
+            val = LSymbol(p.val)
+            if val in (:p, :n)
+                name = :devtype
+                mosfet_type = val == :p ? :pmos : :nmos
+                continue
+            end
+            # Default handling - no rewrite
+        elseif name == :level
+            # TODO
+            level = parse(Float64, String(p.val))
+            continue
+        elseif name == :version
+            version = parse(Float64, String(p.val))
+            continue
+        end
+        val = cg_expr!(state, p.val)
+        push!(params, Expr(:kw, name, Expr(:call, CedarSim.mknondefault, val)))
+    end
+
+    # some devices have a version parameter
+    # while others have distinct models
+    if version !== nothing && modelref.name in (:bsim4,)
+        push!(params, Expr(:kw, :version, Expr(:call, CedarSim.mknondefault, version)))
+    end
+
+    if mosfet_type !== nothing
+        param = devtype_param(modelref, mosfet_type)
+        param !== nothing && push!(params, Expr(:kw, param[1], Expr(:call, CedarSim.mknondefault, param[2])))
+    end
+
+    m = match(binning_rx, LString(model.name))
+    if m !== nothing
+        push!(get!(bins, Symbol(m.captures[1]), Vector{Symbol}()),
+            LSymbol(model.name))
+    end
+
+    lhs = Symbol(LString(model.name))
+
+    @assert isdefined(modelref.mod, modelref.name)
+    T = getglobal(modelref.mod, modelref.name)
+
+    # Peek at the fieldnames of the model to find the correct case for all params.
+    # This obviates the need for using `spicecall`, saving us a bunch of compilation work.
+    # Essentially, we're inlining the generator for case_adjust_kwargs
+    case_insensitive = Dict(Symbol(lowercase(String(kw))) => kw for kw in fieldnames(T))
```

Comment:
Does this mean that all SPICE dataflow is scalar now, except for packing/unpacking for argument passing?

If so, that's great - `Base.merge` was a real source of headaches / compile-time

Reply:
No, `spicecall` is still there for instances, but those usually have fewer parameters, so it generates less code.